README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# phonenumber <img src="man/figures/logo.png" align="right" height="139" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/scumdogsteev/phonenumber/workflows/R-CMD-check/badge.svg)](https://github.com/scumdogsteev/phonenumber/actions)[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/phonenumber)](https://cran.r-project.org/package=phonenumber)
[![Coverage
Status](https://coveralls.io/repos/scumdogsteev/phonenumber/badge.svg?branch=master&service=github)](https://coveralls.io/github/scumdogsteev/phonenumber?branch=master)
[![RStudio CRAN mirror
downloads](https://cranlogs.r-pkg.org/badges/grand-total/phonenumber)](https://cran.r-project.org/package=phonenumber)
<!-- badges: end -->

### Background

When I recently posted some of my [Turbo Pascal
Stuff](https://stevemyles.site/blog/2015/06/19/turbo-pascal-stuff/), I
found an incomplete program that was supposed to do this. I was active
on [BBSes](https://en.wikipedia.org/wiki/Bulletin_board_system) and,
though I don’t recall the reason, I wanted a way to determine the
possible words spelled by the BBS phone numbers (and/or how to determine
what phone numbers correspond to words/phrases). I never got around to
finishing the second part (numbers to letters) in Turbo Pascal, though.

I decided to create this functionality in R for three reasons:

1.  to see if I could write the functions
2.  to learn to publish a package to CRAN
3.  to serve as a possible pedagogical example for others as it involves
    working with lists, splitting strings, and the
    [expand.grid](https://stat.ethz.ch/R-manual/R-devel/library/base/html/expand.grid.html)
    function.

#### Telephone keypad

For purposes of this package, the mapping of numbers to letters on a
telephone’s keypad are as follows:

-   *Default behavior* - if parameter `qz` is omitted (or has a value
    other than 0):
    -   2 corresponds to A, B, C
    -   3 corresponds to D, E, F
    -   4 corresponds to G, H, I
    -   5 corresponds to J, K, L
    -   6 corresponds to M, N, O
    -   7 corresponds to P, Q, R, S
    -   8 corresponds to T, U, V
    -   9 corresponds to W, X, Y, Z
    -   0 and 1 have no corresponding letters
-   *Alternate behavior* - if parameter `qz` = 0:
    -   2 corresponds to A, B, C
    -   3 corresponds to D, E, F
    -   4 corresponds to G, H, I
    -   5 corresponds to J, K, L
    -   6 corresponds to M, N, O
    -   7 corresponds to P, R, S
    -   8 corresponds to T, U, V
    -   9 corresponds to W, X, Y
    -   0 corresponds to Q, Z
    -   1 has no corresponding letters

### Installation

-   `phonenumber` is available [on
    CRAN](https://cran.r-project.org/package=phonenumber) and can be
    installed accordingly:

``` r
install.packages("phonenumber")
library(phonenumber)
```

-   You can also install `phonenumber` from GitHub using the `devtools`
    package:

``` r
install.packages("devtools")
library(devtools)
install_github("scumdogsteev/phonenumber")
library(phonenumber)
```

### Usage

The package consists of two functions:

1.  `letterToNumber` - converts letters in a string to numbers
2.  `numberToLetter` - converts numbers in a string to letters

Both functions convert non-alphanumeric characters to dash (-) and
perform no conversion on their respective base character type (i.e.,
`letterToNumber` leaves letters as is and `numberToLetter` leaves
numbers as is).

### Examples

**`letterToNumber`** converts a string containing letters into the
corresponding numbers on a telephone’s keypad. For example, if the user
wants to know what telephone number corresponds to “Texas:”

``` r
string <- "Texas"
letterToNumber(string)
#> [1] "83927"
```

**`numberToLetter`** converts a string containing numbers into the
corresponding letters on a telephone’s keypad. For example, if the user
wants to know what possible character strings could be spelled by a
sequence of numbers (e.g., 22):

``` r
string <- "22"
numberToLetter(string)
#> [1] "AA" "AB" "AC" "BA" "BB" "BC" "CA" "CB" "CC"
```


NEWS.md
# phonenumber 0.2.3 - DN38416

- add rmarkdown as a dependency for vignette
- update project and other URLs

# phonenumber 0.2.2 - machine phase matter

- tweaks to vignette and documentation

# phonenumber 0.2.1 - phased plasma rifle in the 40-watt range

- change for loops to lapply
- remove ineffective error-checking

# phonenumber 0.2.0 - memetic polyalloy

- add user determination of whether to map "Q" and "Z" to 7 and 9 or to 0
- add additional tests due to new functionality
- add comments to existing examples
- add additional examples due to new functionality
- add package help file
- fix typos and add title to vignette
- add URLs to DESCRIPTION
- updates to README

# phonenumber 0.1.0 - hyperalloy combat chassis

- first version of phonenumber


tests/testthat/test-letterToNumber.R
library(phonenumber)
context("letterToNumber")

test_that("letterToNumber returns the correct numbers", {
    expect_equal(letterToNumber("a"), "2")
    expect_equal(letterToNumber("b"), "2")
    expect_equal(letterToNumber("c"), "2")
    expect_equal(letterToNumber("d"), "3")
    expect_equal(letterToNumber("e"), "3")
    expect_equal(letterToNumber("f"), "3")
    expect_equal(letterToNumber("g"), "4")
    expect_equal(letterToNumber("h"), "4")
    expect_equal(letterToNumber("i"), "4")
    expect_equal(letterToNumber("j"), "5")
    expect_equal(letterToNumber("k"), "5")
    expect_equal(letterToNumber("l"), "5")
    expect_equal(letterToNumber("m"), "6")
    expect_equal(letterToNumber("n"), "6")
    expect_equal(letterToNumber("o"), "6")
    expect_equal(letterToNumber("p"), "7")
    expect_equal(letterToNumber("q"), "7")
    expect_equal(letterToNumber("r"), "7")
    expect_equal(letterToNumber("s"), "7")
    expect_equal(letterToNumber("t"), "8")
    expect_equal(letterToNumber("u"), "8")
    expect_equal(letterToNumber("v"), "8")
    expect_equal(letterToNumber("w"), "9")
    expect_equal(letterToNumber("x"), "9")
    expect_equal(letterToNumber("y"), "9")
    expect_equal(letterToNumber("z"), "9")
    expect_equal(letterToNumber("A"), "2")
    expect_equal(letterToNumber("B"), "2")
    expect_equal(letterToNumber("C"), "2")
    expect_equal(letterToNumber("D"), "3")
    expect_equal(letterToNumber("E"), "3")
    expect_equal(letterToNumber("F"), "3")
    expect_equal(letterToNumber("G"), "4")
    expect_equal(letterToNumber("H"), "4")
    expect_equal(letterToNumber("I"), "4")
    expect_equal(letterToNumber("J"), "5")
    expect_equal(letterToNumber("K"), "5")
    expect_equal(letterToNumber("L"), "5")
    expect_equal(letterToNumber("M"), "6")
    expect_equal(letterToNumber("N"), "6")
    expect_equal(letterToNumber("O"), "6")
    expect_equal(letterToNumber("P"), "7")
    expect_equal(letterToNumber("Q"), "7")
    expect_equal(letterToNumber("R"), "7")
    expect_equal(letterToNumber("S"), "7")
    expect_equal(letterToNumber("T"), "8")
    expect_equal(letterToNumber("U"), "8")
    expect_equal(letterToNumber("V"), "8")
    expect_equal(letterToNumber("W"), "9")
    expect_equal(letterToNumber("X"), "9")
    expect_equal(letterToNumber("Y"), "9")
    expect_equal(letterToNumber("Z"), "9")
    expect_equal(letterToNumber("1"), "1")
    expect_equal(letterToNumber("0"), "0")    
    expect_equal(letterToNumber("-"), "-")
    expect_equal(letterToNumber("#"), "-")
    expect_equal(letterToNumber("!"), "-")
    expect_equal(letterToNumber(" "), "-")
    expect_equal(letterToNumber("adZ 123"), "239-123")    
})

tests/testthat/test-numberToLetter.R
library(phonenumber)
context("numberToLetter")

test_that("numberToLetter returns the correct letters", {
    expect_equal(numberToLetter("2"), c("A", "B", "C"))
    expect_equal(numberToLetter("3"), c("D", "E", "F"))
    expect_equal(numberToLetter("4"), c("G", "H", "I"))
    expect_equal(numberToLetter("5"), c("J", "K", "L"))
    expect_equal(numberToLetter("6"), c("M", "N", "O"))
    expect_equal(numberToLetter("7"), c("P", "Q", "R", "S"))
    expect_equal(numberToLetter("7", qz = 0), c("P", "R", "S"))
    expect_equal(numberToLetter("8"), c("T", "U", "V"))
    expect_equal(numberToLetter("9"), c("W", "X", "Y", "Z"))
    expect_equal(numberToLetter("9", qz = 0), c("W", "X", "Y"))
    expect_equal(numberToLetter("1"), "1")
    expect_equal(numberToLetter("0"), "0")
    expect_equal(numberToLetter("0", qz = 0), c("Q", "Z"))
    expect_equal(numberToLetter("-"), "-")
    expect_equal(numberToLetter("#"), "-")
    expect_equal(numberToLetter("!"), "-")
    expect_equal(numberToLetter(" "), "-")
    expect_equal(numberToLetter("911!"), c("W11-", "X11-", "Y11-", "Z11-"))    
    expect_equal(numberToLetter("000"), "000")
    expect_equal(numberToLetter("000", qz = 0), c("QQQ", "QQZ", "QZQ", "QZZ", 
                                                  "ZQQ", "ZQZ", "ZZQ", "ZZZ"))
})

R/letterToNumber.R
#' 
#' Convert letters to numbers as on a telephone's keypad
#' 
#' Take a character vector and convert it to the equivalent number sequence
#' from a telephone's key pad
#' 
#' @param value An input value as a character vector with one element (a string)
#' @param qz Whether to assign q and z to zero (qz = 0) or not (any other value)
#' 
#' @export
#' 
#' @examples
#' # Convert an alphabetic string can be converted directly (with 
#' # non-alphanumeric characters replaced by dashes)
#' 
#' letterToNumber("R functions") # returns "7-386284667"
#' 
#' # Of course, vectors containing strings can also be converted
#' 
#' string <- "Phone Number"
#' letterToNumber(string) # returns "74663-686237"
#' 
#' # Alphanumeric strings can also be converted with numbers being returned as
#' # is
#' 
#' letterToNumber("Jenny's Number is 867-5309") # returns "53669-7-686237-47-867-5309"
#' 
#' # Specifying qz = 0 maps "q" and "z" to 0 instead of 7 and 9
#' 
#' letterToNumber("qz") # returns "79"
#' letterToNumber("qz", qz = 0) # returns ("00")
#' 
#' @return A character vector of numbers and dashes based on value

letterToNumber <- function(value, qz = 1) {
    value <- as.character(value)
    value <- gsub("[^A-Za-z0-9]", "-", value)
    value <- toupper(value)
    valueSplit <- strsplit(value, "")[[1]]
    numString <- as.character()
    mapphone <- function(char) { 
        if (qz == 0 && (char == LETTERS[17] | char == LETTERS[26])) { "0" } 
        else { ifelse(is.element(char, LETTERS[1:3]), "2",
               ifelse(is.element(char, LETTERS[4:6]), "3",
               ifelse(is.element(char, LETTERS[7:9]), "4",
               ifelse(is.element(char, LETTERS[10:12]), "5",
               ifelse(is.element(char, LETTERS[13:15]), "6",
               ifelse(is.element(char, LETTERS[16:19]), "7",
               ifelse(is.element(char, LETTERS[20:22]), "8",
               ifelse(is.element(char, "-") | 
                      suppressWarnings(!is.na(as.numeric(char))), char, "9")
               ))))))) 
        } 
        } 
    numString <- lapply(valueSplit, mapphone)
    return(paste0(numString, collapse = ""))
}

R/numberToLetter.R
#' 
#' Convert numbers to letters as on a telephone's keypad
#' 
#' Take a character vector (i.e., a telephone number) and convert it to all
#' all possible letter combinations as on from a telephone's key pad
#' 
#' @param value An input value as a character vector with one element (a string)
#' @param decreasing Whether to sort the results in alphabetical order or not
#' @param qz Whether to assign q and z to zero (qz = 0) or not (any other value)
#' 
#' @export
#' 
#' @examples
#' # Convert a string or a vector of numeric characters
#'
#' numberToLetter("911") # returns "W11" "X11" "Y11" "Z11"
#' x <- "911"
#' numberToLetter(x) # also returns "W11" "X11" "Y11" "Z11"
#' 
#' # Convert a number directly
#' 
#' numberToLetter(911) # also returns "W11" "X11" "Y11" "Z11"
#' 
#' # Convert an alphanumeric string (letters are returned as is and
#' # non-alphanumeric characters are returned as dashes)
#' 
#' numberToLetter("dial 911!") # returns "DIAL-W11-" "DIAL-X11-" "DIAL-Y11-" "DIAL-Z11-"
#'
#' # Specifying qz = 0 maps "q" and "z" to 0 instead of 7 and 9
#' 
#' numberToLetter("000") # returns "000" 
#' numberToLetter("000", qz = 0) # returns "QQQ" "QQZ" "QZQ" "QZZ" "ZQQ" "ZQZ" "ZZQ" "ZZZ"
#' 
#' @return A character vector of letters and dashes based on value

numberToLetter <- function(value, decreasing = FALSE, qz = 1) {
    if(is.na(value)) { stop("Argument must be a string.") } else {
        value <- as.character(value)
        value <- gsub("[^A-Za-z0-9]", "-", value)
        value <- toupper(value)
        valueSplit <- strsplit(value, "")[[1]]    
        strList <- list()
        phonemap <- function(char) {
            if (char == "2") { strsplit(LETTERS[1:3], "") } else {
            if (char == "3") { strsplit(LETTERS[4:6], "") } else {
            if (char == "4") { strsplit(LETTERS[7:9], "") } else {
            if (char == "5") { strsplit(LETTERS[10:12], "") } else {
            if (char == "6") { strsplit(LETTERS[13:15], "") } else {
            if (char == "7") { strsplit(LETTERS[16:19], "") } else {
            if (char == "8") { strsplit(LETTERS[20:22], "") } else {
            if (char == "9") { strsplit(LETTERS[23:26], "") } else { char } 
            }}}}}}}
        }
        for (char in valueSplit) {        
            letterSplit <- if (qz == 0) { if (char == "0") { list(LETTERS[17], 
                                                                  LETTERS[26]) 
                                        } else { 
                                          if (char == "7") { c(LETTERS[16], 
                                                      strsplit(LETTERS[18:19],
                                                               "")) } else { 
                                          if (char == "9") { 
                                              strsplit(LETTERS[23:25], "") }
                                          else { phonemap(char) }}}} 
                           else { phonemap(char) }
            strList[[length(strList) + 1]] <- letterSplit
        }
        strDF <- expand.grid(strList)
        lettString <- do.call(paste0, strDF[1:ncol(strDF)])
        lettString <- sort(lettString, decreasing)
        return(lettString)
    }
}